dependencies/http/http_test.go

```diff
@@ -54,23 +57,88 @@ func TestLimitedDefaultClient(t *testing.T) {
 	})
 }
 
-// The TestValidator will let everything through _except_ a specific url, which
-// is redirected to from a http test server.
+// Mocking DNS to resolve to blocked IP ranges and ensuring the connections
+// fail. ADT: Would be nice to also test some pass cases, but it is unclear how
+// to prevent the actual network connection out to the public internet. It's
+// the dialer control function we are testing, so we can't mock the dialer to
+// achieve that. Can really only replace the validator to pass a connetion to
+// localhost. Doing that below the redirect test.
+func TestIpValidation(t *testing.T) {
+	bad := map[string]mockdns.Zone{
+		"bad01.example.org.": {A: []string{"127.0.0.1"}},
+		"bad02.example.org.": {A: []string{"127.10.20.30"}},
+		"bad03.example.org.": {A: []string{"172.23.1.2"}},
+		"bad04.example.org.": {A: []string{"192.168.0.0"}},
+		"bad05.example.org.": {A: []string{"169.254.0.1"}},
+		"bad06.example.org.": {AAAA: []string{"0000:0000:0000:0000:0000:0000:0000:0001"}},
+		"bad07.example.org.": {AAAA: []string{"fc80:0000:0000:0000:0000:0000:0000:0001"}},
+		"bad08.example.org.": {AAAA: []string{"fc00:0000:0000:0000:0000:0000:0000:0001"}},
+	}
+
+	// Mock a dns server.
+	dnsLogger := log.New(ioutil.Discard, "mockdns server: ", log.LstdFlags)
+	srv, _ := mockdns.NewServerWithLogger(bad, dnsLogger, false)
+	defer srv.Close()
+
+	// Hack the default resolver to use this server.
+	srv.PatchNet(net.DefaultResolver)
+	defer mockdns.UnpatchNet(net.DefaultResolver)
+
+	client := NewDefaultClient(depsUrl.PrivateIPValidator{})
+
+	for k := range bad {
+		req, err := http.NewRequest("POST", ("http://" + k + "/path"), bytes.NewReader([]byte{}))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		_, err = client.Do(req)
+		if err == nil {
+			t.Fatal("expected private IP validation error, but client.Do succeeded")
+		} else {
+			if !strings.HasSuffix(err.Error(), "it connects to a private IP") {
+				t.Fatalf("expected private IP validation error, but got %v", err)
+			}
+		}
+	}
+}
+
+// The TestValidator will let everything through _except_ a specific IP, which
+// is redirected to from an http test server. We must mock the validator
+// because the original connection goes to localhost and that must be
+// permitted, so it can send the redirect.
 type TestValidator struct{}
 
 func (TestValidator) Validate(anUrl *url.URL) error {
-	if anUrl.Host == "test-validator.example.com" {
+	return nil
+}
+
+func (TestValidator) ValidateIP(ip net.IP) error {
+	if ip.Equal(net.ParseIP("127.6.6.6")) {
 		return errors.New(codes.Invalid, "url validation error, it connects to a private IP")
 	}
 	return nil
 }
 
 func TestInvalidRedirects(t *testing.T) {
+	bad := map[string]mockdns.Zone{
+		"bad01.example.com.": {A: []string{"127.6.6.6"}},
+	}
+
+	// Mock a dns server.
+	dnsLogger := log.New(ioutil.Discard, "mockdns server: ", log.LstdFlags)
+	srv, _ := mockdns.NewServerWithLogger(bad, dnsLogger, false)
```

Comment:
This is neat. I'd like to take a look at this, as this is how I usually expect mocking to occur, rather than the code generation most mock libraries seem to use.

Reply:
Yeah dug this up while looking for DNS resolution mocking and it was the only approach I found. Basically starts a DNS server on a random port and then alters the global default `Resolver` struct to connect to that for DNS, instead of the system settings. Maybe a bit heavy but couldn't find a way to hook the DNS lookup in go.